greedy_algorithms/job_sequencing_algorithm.cpp

```diff
@@ -0,0 +1,197 @@
+ * -------|-----------|--------
+ *   a    |     2     |  100
+ *   b    |     1     |  19
+ *   c    |     2     |  27
+ *   d    |     1     |  25
+ *   e    |     3     |  15
+ * Output: Following is maximum profit sequence of jobs: c, a, e
+ *
+ *
+ *
+ * Naive Approach:
+ * Generate all subsets of a given set of jobs and check individual subsets for
+ * the feasibility of jobs in that subset. Keep track of maximum profit among
+ * all feasible subsets.
+ * The above approach's time complexity is: O(N^2) with auxliary space O(N)
+ * N refers to the number of jobs.
+ *
+ *
+ * Greedy Approach:
+ * Greedily choose the jobs with maximum profit first, by sorting the jobs in
+ * decreasing order of their profit. This would help to maximize the total
+ * profit as choosing the job with maximum profit for every time slot will
+ * eventually maximize the total profit.
+ * The above approaches's time complexity is: O(NlogN) with auxliary space o(N)
+ * N refers to the number of jobs.
+ *
+ *
+ * Greedy approach psuedo-code:
+  - Sort all jobs in decreasing order of profit.
+  - Iterate on jobs in decreasing order of profit.
+       - For each job , do the following:
+           - Find a time slot i, such that slot is empty and i < deadline and i
+             is greatest.
+           - Put the job in this slot and mark this slot filled.
+           - If no such i exists, then ignore the job.
+ *
+ * @author [Mahmoud Elkholy](https://github.com/maahmoudezzat)
+ */
+
+// Implementation of the greedy algorithm using priority_queue AKA max_heap.
+// C++ code for the above approach
+#include <algorithm>  // for sort function.
+#include <cassert>    // for assert funtion in testing
+#include <iostream>   // for standard in/out
+#include <queue>      // for priority queue data structure
+#include <vector>     // for vector data structure
+
+/**
+ * @namespace
+ * @brief Greedy Algorithms
+ */
+namespace greedy_algorithms {
+
+/**
+ * @brief A structure to represent a job
+ */
+struct Job {
+    char id;     // Job Id
+    int dead;    // Deadline of job
+    int profit;  // Profit earned if job is completed before deadline
+};  // namespace greedy_algorithmsstruct Job
+
+// Custom sorting helper struct which is used for sorting
+// all jobs according to profit
+
+/**
+ * @brief Utility function that finds Custom sorting helper
+ * struct which is used for sorting all jobs according to profit
+ *
+ * @param a first job struct
+ * @param b second job struct
+ * @returns true if the proft of a less than profit of b, else false.
+ */
+struct jobProfit {
+    bool operator()(Job const& a, Job const& b) {
+        return (a.profit < b.profit);
+    }
+};
+
+/**
+ * @brief function that get the maximum profit from jobs
+ *
+ * @param arr array fo Jobs
+ * @param n number of the jobs in the array
+ * @returns vector of 'char' including the IDs of jobs
+ * that has the maximum profit from jobs
+ *
+ */
+std::vector<char> getJobScheduling(Job arr[], int n) {
```

Comment:
cant arr be a vector or even array since it doesn't seem to be expanded during runtime.

Reply:
> You're right it is not expanding during the runtime, so that's why we use arr.

use std::array instead, its better

Reply:
Good point, Its better in performance but sometimes it looks scarier for learners since they're start learning with built-in arrays but I have modified it so they can get in touch with std::array.